### DISCUSSION TITLE
Using TPE Sampler: How to know the cut-off between random and TPE?

### DISCUSSION THREAD

**Discussion by ilenoa** [optuna/optuna#3969] [Q&A] [answered]
Hello, I am trying to use TPESampler and I am having difficulties identifying when TPE sampling is being done, rather than the initial random sampling. I am checking the CSV with all trials obtained using the function _study.trials_dataframe_, but there are no system attributes of the trials that help with the identification. I have a few questions to be able to do it on my own:

1) Is it possible to identify the sampling done per trial?
2) Let's say n_startup_trials is 100. Does this mean TPE will start sampling on trial 101, or does it need 100 completed (not pruned) trials?
3) Let's say that again n_startup_trials is 100. I start a new study with a DB, reach 80 trials and then stop. Then, I start it again, with the same parameters, therefore my first trial would be 81. Will optuna wait until trial 181 to do TPE, or will it start it at 101 with the trials already saved on the database?
4) Let's say I start a new study and use GridSampler and run 200 trials, also with a DB. If I then stop and start the same study with TPE Sampler and n_startup_trials is 100, will it use the trials of GridSampler as random (and therefore start TPE at trial 201), or will it do 100 random and then TPE at trial 301?
5) I have noticed that TPE tends to repeat the same trials over and over. I saw other comments addressing this issue too, and it was recommended to increase the parameter n_startup_trials. My question is: If I run TPE for 1K trials with n_startup_trials = 100, check it out, and see that I want it to explore more, could I run TPE again (with the same study name on the DB) for another 1K trials with n_startup_trials=100, so it does random trials 1000-1100, and then TPE? And if so, will this second round of TPE use the previous 1100 trials to make a guess, or just the last 100?

Thank you in advance for your time and answers.

**Comment by nzw0301** [comment, marked as answer]
> Is it possible to identify the sampling done per trial?

No. But if we access `n_startup_trials`, we can infer the current sampler from the `trial.number` and store the sampler name as [a user attribute](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/003_attributes.html).

> Let's say n_startup_trials is 100. Does this mean TPE will start sampling on trial 101

Yes, more precisely, from a trial whose `trial.number` is `100` since Optuna's trial number is zero-origin.

> does it need 100 completed (not pruned) trials?

No, complete and pruned trials are always included for `n_startup_trials`. Depending on `constant_liar` argument, it also cares trials whose state is `RUNNING` by https://github.com/optuna/optuna/blob/2ab292f7fb4b432d15ef793c4db42d2358eca51e/optuna/samplers/_tpe/sampler.py#L623-L627.

> Will optuna wait until trial 181 to do TPE, or will it start it at 101 with the trials already saved on the database?

TPE starts at the 101-th trial since complete and pruned trials are on the database. 

> Let's say I start a new study and use GridSampler and run 200 trials, also with a DB. If I then stop and start the same study with TPE Sampler and n_startup_trials is 100, will it use the trials of GridSampler as random (and therefore start TPE at trial 201), or will it do 100 random and then TPE at trial 301?

This case: "will it use the trials of GridSampler as random (and therefore start TPE at trial 201)".

> My question is: If I run TPE for 1K trials with n_startup_trials = 100, check it out, and see that I want it to explore more, could I run TPE again (with the same study name on the DB) for another 1K trials with n_startup_trials=100, so it does random trials 1000-1100, and then TPE?

In this case, TPE will be used for trials 1000 -- 1100 if we use the same `study`.

**Reply by ilenoa** [reply]
Thank you for the detailed response. Therefore, if one wanted to force some exploration after TPESampler, the way to do it would be to stop TPE Sampler, run Random Sampler for a while and then run again TPE Sampler, right?

**Reply by nzw0301** [reply]
Yeah, it sounds good.
